lgc/util/ElfNoteEntryInsertionUtil.cpp

```diff
@@ -0,0 +1,258 @@
+    lengthToBeShifted = newOffset - sectionHeader.sh_offset;
+
+    // Update the offset of section pointed by the section header to its new offset.
+    sectionHeader.sh_offset = newOffset;
+  }
+
+  // Sort sectionAndNewOffset by the new offset of each section in the increasing order.
+  sort(sectionAndNewOffset,
+       [](const SectionShiftInfo &i0, const SectionShiftInfo &i1) { return i0.newOffset < i1.newOffset; });
+  return sectionAndNewOffset;
+}
+
+// =====================================================================================================================
+// Inserts the new contents to the given ELF.
+//
+// @param [in/out] elf        : The ELF to insert the new contents.
+// @param insertionOffset     : The offset of the new contents to be inserted.
+// @param elfContentStream    : The ELF contents to be inserted.
+// @param sectionAndNewOffset : The sections to be shifted and their new offsets.
+void insertContentsToELF(SmallVectorImpl<char> &elf, const ELF::Elf64_Off insertionOffset,
+                         AppendingBinaryByteStream &elfContentStream,
+                         const SmallVectorImpl<SectionShiftInfo> &sectionAndNewOffset) {
+  // Strip sections after the insertion offset of the new contents.
+  elf.resize(insertionOffset);
+
+  // Write the new contents.
+  raw_svector_ostream elfStream(elf);
+  ArrayRef<uint8_t> noteEntry;
+  auto error = elfContentStream.readBytes(0, elfContentStream.getLength(), noteEntry);
+  (void)error;
+  assert(!error);
+  elfStream << StringRef(reinterpret_cast<const char *>(noteEntry.data()), noteEntry.size());
+
+  // Write the sections after the insertion offset.
+  for (const auto &sectionAndNewOffsetInfo : sectionAndNewOffset) {
+    elfStream.write_zeros(sectionAndNewOffsetInfo.newOffset - elfStream.str().size());
+    elfStream << sectionAndNewOffsetInfo.section.str();
+  }
+}
+
+// =====================================================================================================================
+// Write the section header table to the given offset.
+//
+// @param [in/out] elf             : The ELF to write the section header table.
+// @param sectionHeaderTableOffset : The offset where it will write the section header table.
+// @param sectionHeaderTable       : The section header table to be written to the ELF.
+void writeSectionHeaderTable(SmallVectorImpl<char> &elf, const ELF::Elf64_Off sectionHeaderTableOffset,
+                             const SmallVectorImpl<ELF::Elf64_Shdr> &sectionHeaderTable) {
+  if (sectionHeaderTable.size() == 0)
+    return;
+
+  raw_svector_ostream elfStream(elf);
+  unsigned minElfSizeForSectionHeaders = sectionHeaderTableOffset + sizeof(ELF::Elf64_Shdr) * sectionHeaderTable.size();
+  if (minElfSizeForSectionHeaders > elf.size())
+    elfStream.write_zeros(minElfSizeForSectionHeaders - elf.size());
+  elfStream.pwrite(reinterpret_cast<const char *>(sectionHeaderTable.data()),
```

Comment:
Consider creating a helper function that takes `SmallVectorImpl<T>` and returns a `StringRef`. It seems a bit bugprone to me if you have to remember to consider the element size in every callsite when you do this conversion.
This helper function could also take a starting offset argument, since you seem to need it in a few places.